Add implementation of multi queue worker proposal

Ayiga opened the pull request:
<!-- <!-- Closes #<ISSUE_NUMBER> -->
<!-- These comments should help create a useful PR message, please delete any remaining comments before opening the PR. -->
<!-- If there is no issue number make sure to describe clearly *why* this PR is necessary. -->
<!-- Mention open questions, remaining TODOs, if any -->

### This PR:
<!-- Describe what this PR adds to this repo and why -->
<!-- E.g. -->
<!-- * Implements feature 1 -->
<!-- * Fixes bug 3 -->
Adds an initial implementation of the [proposal to improve this submission process](https://www.notion.so/espressosys/Nitro-Integration-Submit-to-Espresso-Process-Change-Proposal-2372431b68e98006b618e82adcb0f0b3).

This is done by separating the responsbilities of submitting to the Espresso Network and verifying the presence of the resulting Transaction on the Espresso Network as discrete untis of work.  These pieces of work can then be delivered and assigned to workers who are only dedicated and focused on the processing of these discrete units of work.  This is done via worker queues with a managed scheduler.  This approach / implementation of worker queues is based on an article to take advantage of [channels of channels](https://nesv.ca/posts/worker-queues-in-go/) as there is evidence of improved throughput.

Depends on #692 

<!-- ### This PR does not: -->
<!-- Describe what is out of scope for this PR, if applicable. Leave this section blank if it's not applicable -->
<!-- This section helps avoid the reviewer having to needlessly point out missing parts -->
<!-- * Implement feature 3 because that feature is blocked by Issue 4 -->
<!-- * Implement xyz because that is tracked in issue #123. -->
<!-- * Address xzy for which I opened issue #456 -->

<!-- ### Key places to review: -->
<!-- Describe key places for reviewers to pay close attention to -->
<!-- * file.rs, `add_integers` function -->
<!-- Or directly comment on those files/lines to make it easier for the reviewers -->

<!-- ### How to test this PR:  -->
<!-- Optional, uncomment the above line if this is relevant to your PR -->
<!-- If your PR is fully tested through CI there is no need to add this section -->
<!-- * E.g. `just test` -->

<!-- ### Things tested -->
<!-- Anything that was manually tested (that is not tested in CI). -->
<!-- E.g. building/running of docker containers. Changes to docker demo, ... -->
<!-- Especially mention anything untested, with reasoning and link an issue to resolve this. -->

<!-- Complete the following items before creating this PR -->
<!-- [ ] Issue linked or PR description mentions why this change is necessary. -->
<!-- [ ] PR description is clear enough for reviewers. -->
<!-- [ ] Documentation for changes (additions) has been updated (added).  -->
<!-- [ ] If this is a draft it is marked as "draft".  -->
<!-- [ ] For integration projects, make sure it won't break backward compatibility. -->

<!-- Details on maintaining backward compatibility for integration projects: -->
<!-- * Try to keep changes additive: add new, optional methods, flags, or parameters instead of modifying or removing existing functionality. -->
<!-- * If modification is necessary, it should be either a clear bug fix or guarded by a config/feature flag.  -->
<!-- * Follow Open Closed Principle and Interface Segregation Principle, clients should not be forced to depend on interfaces they do not use. -->

<!-- To make changes to this template edit https://github.com/EspressoSystems/.github/blob/main/PULL_REQUEST_TEMPLATE.md -->


---
- To see the specific tasks where the Asana app for GitHub is being used, see below:
  - https://app.asana.com/0/0/1210850486591824
  - https://app.asana.com/0/0/1210850486591818
  - https://app.asana.com/0/0/1210850486591826

Ayiga committed:
Revise TestTransactionStreamerEspressoThroughput conditions

The TestTransactionStreamerEspressoThroughput test originally was created just to
monitor throughput based on the parameters being configured for the mock
setups.  The idea was to determine how interval configuration impacted the integration
performance under idealized conditions.

With some base lines set, it is now important to try and ensure that our integration
continues to perform at an established baseline and well beyond in order to ensure
that we ware able to catch up in a sufficient amount of time, should we fall behind.

To achieve this goal the TestTransactionStreamerEspressoThroughput test has been
reworked to test specific scenarios against the required benchmarks.  Additionally
it is important to codify the requirements of it being able to outperform the baseline
by a factor.  In this case the factor chosen is around 8x, but this was chosen arbitrarily
and our tests have indicated that we are able to exceed this with the right polling
interval configuration.

Ayiga committed:
Add Multi Worker Queue Espresso Submitter

Based on potential issues identified in the polling espresso submitter, an alternative
approach has been proposed.  This approach focuses on performing different
parts of submitting transactions to espresso in discrete units of work that can be
scaled independently.  This would allow for multiple Espresso Transactions to be
inflight in terms of submission and verification at the same time, in order to maximize
the amount of throughput that is allowed.

To realize this approach this adds the `MultiWorkerQueueEspressoSubmitter` which
manages and runs two separate work queues for submitting and verifying transactions
that are submitted to Espresso.  The idea here is that this will only focus on the parts
related to Espresso as much as possible without needing to worry about the parts
that are specific to the underlying integration.

By taking this approach, the result should be a more reliable and trackable process for
submitting and verifying transactions to the Espresso Network.

NOTE: There is still work to be done here with regards to adapting the specific
integration to the transactions being submitted to Espresso, and retrieving them
from Espresso.

This also updates the `TestTransactionStreamerEspressoThroughput` test to measure
the throughput performance with targets as compared to the Polling Espresso
Submitter solution.

zacshowa commented on espresso/submitter/multi_worker_queue_espresso_submitter.go:
```diff
@@ -0,0 +1,1012 @@
+)
+
+// MultiWorkerQueueEspressoSubmitter is an implementation of `EspressoSubmitter`
+// that utilities multiple worker queues to perform Transaction Submission to
```
```suggestion
// that utilizes multiple worker queues to perform Transaction Submission to
```

Ayiga replied:
Addressed in 34ec8833dd95e6bd77166426858fde7b66e51abf.

Ayiga committed:
Separate the Nitro Message handling from Espresso Transactions

In order to make the MultiWorkerQueueEspressoSubmitter more agnostic and to not
have to be associated with the details of the Nitro network, it is beneficial to have
a clear boundary of separation where possible.

To facilitate this, the implementation has been separated into two different managing
structures, one that manages the Espresso Transaction Submission process in
isolation, and one that handles the Nitro Message submission, bundling, and
transaction submission scheduling.

Ayiga committed:
Remove magic values in favor of configurable ones

There are a few values in the Multi Worker Queue solution that are set to hard coded
magic values.  In order to favor flexibility and configurability, these values should be
adjustable in some way.

This change adds configuration values and functional options for all of these values
and moves the current default values to the configuration defaults variable.

Sneh1999 commented on espresso/submitter/multi_worker_queue_espresso_submitter.go:
```diff
@@ -0,0 +1,1175 @@
+// convertToUint64WithFallback is a helper function that converts a signed
+// integer value to a type whose underlying type is an unsigned 64-bit integer,
+// and returns a fallback value if the signed value is negative. If the value
+func convertToUint64WithFallback[T signedInteger, U ~uint64](value T, fallback U) U {
```
should we create a utils file for the submitter and add these functions there?

Ayiga replied:
Done in b59e20dd78d76f35b20deb5ea403d6eb9a10a0f7.

Sneh1999 commented on espresso/submitter/multi_worker_queue_espresso_submitter.go:
```diff
@@ -0,0 +1,1175 @@
+func (n *NitroMessageToEspressoTransactionAdapter) bundleTransactions(startPos, endPos arbutil.MessageIndex) (arbutil.MessageIndex, error) {
+	i := startPos
+	pendingTxnsPos := createSliceOfIntegerForRangeInclusive(startPos, endPos)
+	for i < endPos {
```
should this be `i <= endPos` because `createSliceOfIntegerForRangeInclusive` is inclusive of endPos?

zacshowa replied:
I think this would iterate through all of the messages in list actually.
When we build the hotshot payload, we return the msg count of the last message that was included in the hotshot payload and then convert it to a uint64 and assign that value to `i`. If `i` is ever == `endPos`, then then it means that endPos was included in the last hotshot transaction

Ayiga replied:
I think in most cases this would be covered, however I do think there is an edge case when the `startPos` is equal to the `endPos`.

Addressed in 12043cf25b852a052f56cf23b4dff4b35ecd34b6.

Sneh1999 commented on espresso/submitter/multi_worker_queue_espresso_submitter.go:
```diff
@@ -0,0 +1,1175 @@
+	for {
+		select {
+		case <-ctx.Done():
+			return
```
can we add a log here to figure out this happened in case it does?

Ayiga replied:
Addressed in 690fdf595a55c157fb88cbf2f2429b48a7458990.

Sneh1999 commented on espresso/submitter/multi_worker_queue_espresso_submitter.go:
```diff
@@ -0,0 +1,1175 @@
+
+		case pos, ok := <-n.availableTransaction:
+			if !ok {
+				// The channel is closed, so we can exit
```
same here

Ayiga replied:
Addressed in 690fdf595a55c157fb88cbf2f2429b48a7458990.

Sneh1999 commented on espresso/submitter/multi_worker_queue_espresso_submitter.go:
```diff
@@ -0,0 +1,1175 @@
+				continue
+			}
+
+			submittedPos = nextPos
```
what will happen in case the batcher restarts? Will we lose all our progress because we are storing it in submittedPos in memory same with availablePos?

Ayiga replied:
These positions dictate what attempts to get submitted to Espresso, and what's available.  If we have a new `availablePos`, then that indicates that all messages from `0` through `availablePos` are available to retrieve.  Additionally, the `submittedPos` keeps track of the next position to submit to `Espresso`.  At the time of this comment, these were both defaulting to `0` due to a lack of configurability.

The idea is that the consumer should be able to specify these initial values in order to indicate a starting state.  This configurability has been added in a5b71fed2ea3d720673e82194de163af2d1c9962.

Sneh1999 commented on espresso/submitter/multi_worker_queue_espresso_submitter.go:
```diff
@@ -0,0 +1,1175 @@
+// We retrieve the Attestation using our epheremal public key we created in EspressoKeyManager
+// After we retrieve, we verify the attestation, where we retrieve the result
+// Which will contain the complete attestation which we serialize for further processing
+func (t *NitroMessageToEspressoTransactionAdapter) getNitroAttestation(pubKey []byte) ([]byte, error) {
```
Maybe we can move this attestation code for SGX and Nitro to another file for cleanup purposes and name it attestation_helper.go or something else? Just want to keep the code in this file small

Ayiga replied:
I think all of the Attestation code should probably be removed from the Submitters entirely.  I don't see any reason why they exist within these structures at all.  They seem intrinsically linked to the KeyManager functionality, and are better placed there, but still even better placed in their own area.  For this effort, I'm simply attempting to match the implementation of the other Submitter, and those sort of refactors can be done at a later time less this PR become about doing multiple things.

Sneh1999 commented on espresso/submitter/multi_worker_queue_espresso_submitter.go:
```diff
@@ -0,0 +1,1175 @@
+// `start`, and ending at `endInclusive`, inclusive.
+//
+// Comprehension: [x | x ∈ [start, endInclusive]]
+func createSliceOfIntegerForRangeInclusive[T integer](start, endInclusive T) []T {
```
same lets move `createSliceOfIntegerForRangeInclusive` and `integer` to a utils file

ImJeremyHe replied:
Maybe we don’t need generics here? They’re usually helpful when a function is reused in multiple places or when we want to abstract over complex types for testing, but that doesn’t seem to apply in this case.

Ayiga replied:
Currently it likely doesn't need to be generic, but it could potentially be useful in the future.

Done in b59e20dd78d76f35b20deb5ea403d6eb9a10a0f7.

Sneh1999 commented on espresso/submitter/multi_worker_queue_espresso_submitter.go:
```diff
@@ -0,0 +1,1175 @@
+
+// Error implements error
+func (e ErrorFailedToLaunchTransactionIncludedQueueWorker) Error() string {
+	return fmt.Sprintf("failed to launch transaction included queue worker %d: %v", e.Worker, e.Cause)
```
why not use `fmt.Errof`?

Ayiga replied:
In general, when it comes to `error` handling in languages, and especially in `golang`, there tend to be a few different ways to do them.  One of the absolute worst ways to do them, in my experience, is to simply treat them as `string`s.  When you treat errors as strings, then the only way to check for errors is via string comparison.  This makes the error checking extremely brittle, as if you change the error message for any reason, then the error detection method needs to change.

If instead we opt for specific error types, then the error becomes resilient to these types of changes, and even encourages and allows for adjustments to the underlying error structure to contain more context when it is needed.  There are a couple down-sides to this approach. one is that there is more boiler-plate in terms of defining the structure and the methods, but the benefits gained are well worth the minor effort that that incurs.  The other has to do with serialization.  In order to effectively transmit and recover the error, you need some way to determine which error to deserialize given a `byte` `slice`.  This becomes especially difficult in `golang`'s case where the idiomatic approach to error handling is to just return an `error`.  Deserializing something that is an `error` interface is fraught with many challenges.  However, I find that having specific error structures available is still worth that challenge as well.

To that end, while `fmt.Errorf` is a convenience that is helpful for a quick error when developing something, their presence indicates the existence of brittle, and potentially hard to maintain code when it comes to error inspection.

Sneh1999 commented on espresso/submitter/multi_worker_queue_espresso_submitter.go:
```diff
@@ -0,0 +1,1175 @@
+		// Wait for a job to be sent to us
+		job, ok := <-ch
+		if !ok {
+			log.Info("Submit transaction job queue closed, exiting", "worker", w.id)
```
should we make this a `Warn`?

Ayiga replied:
Addressed in 690fdf595a55c157fb88cbf2f2429b48a7458990.

Sneh1999 commented on espresso/submitter/multi_worker_queue_espresso_submitter.go:
```diff
@@ -0,0 +1,1175 @@
+		if err != nil {
+			// This is unfortunate, but we shouldn't stop processing the job
+			// just because of a small reference failure.
+			log.Info("Failed to estimate transaction hash", "error", err, "commit", common.Hash(commit), "worker", w.id)
```
same here can we make this an error/warn?

Ayiga replied:
Addressed in 690fdf595a55c157fb88cbf2f2429b48a7458990.

Sneh1999 commented on espresso/submitter/multi_worker_queue_espresso_submitter.go:
```diff
@@ -0,0 +1,1175 @@
+		// Process the job
+		if job.attempt > 0 {
+			// Let's slow down a little
+			time.Sleep(convertToInt64WithFallback[uint, time.Duration](job.attempt, 0) * 100 * time.Millisecond)
```
lets make this configurable, so that anyone can choose the slowing speed. Also can you add some comments on why we are doing this here?

Ayiga replied:
Addressed in ec77acd5e4b5dba0af7c53849cf36455d0dd5e53.

Sneh1999 commented on espresso/submitter/multi_worker_queue_espresso_submitter.go:
```diff
@@ -0,0 +1,1175 @@
+		// Wait for a job to be sent to us
+		job, ok := <-ch
+		if !ok {
+			log.Info("Transaction inclusion job queue closed, exiting", "worker", w.id)
```
same here convert to warn/error

Ayiga replied:
Addressed in 690fdf595a55c157fb88cbf2f2429b48a7458990.

Sneh1999 commented on espresso/submitter/multi_worker_queue_espresso_submitter.go:
```diff
@@ -0,0 +1,1175 @@
+		}
+
+		details, err := w.client.FetchTransactionByHash(ctx, &job.txnHash)
+		if err != nil {
```
should we log the error here?

Ayiga replied:
Addressed in 690fdf595a55c157fb88cbf2f2429b48a7458990.

Sneh1999 commented on espresso/submitter/multi_worker_queue_espresso_submitter.go:
```diff
@@ -0,0 +1,1175 @@
+
+		if job.attempt > 0 {
+			// Let's slow down a little
+			time.Sleep(convertToInt64WithFallback[uint, time.Duration](job.attempt, 0) * 100 * time.Millisecond)
```
Should we make this configurable?

Ayiga replied:
Addressed in ec77acd5e4b5dba0af7c53849cf36455d0dd5e53.

Sneh1999 commented on espresso/submitter/multi_worker_queue_espresso_submitter.go:
```diff
@@ -0,0 +1,1175 @@
+	for i := uint64(0); i < w.numSubmitTransactionWorkers; i++ {
+		worker, workerOk := <-w.submitTxnsJobQueue
+		if !workerOk {
+			// The job queue is closed, so we can exit.
```
log the warning in case we need to debug

Ayiga replied:
Addressed in 690fdf595a55c157fb88cbf2f2429b48a7458990.

Sneh1999 commented on espresso/submitter/multi_worker_queue_espresso_submitter.go:
```diff
@@ -0,0 +1,1175 @@
+			// job to it.
+
+			if !ok {
+				log.Info("Submit transaction job queue closed, exiting")
```
change to warn

Ayiga replied:
Addressed in 690fdf595a55c157fb88cbf2f2429b48a7458990.

Sneh1999 commented on espresso/submitter/multi_worker_queue_espresso_submitter.go:
```diff
@@ -0,0 +1,1175 @@
+	for {
+		select {
+		case <-ctx.Done():
+			log.Info("Submit transaction scheduler exiting")
```
change logging level

Ayiga replied:
Addressed in 690fdf595a55c157fb88cbf2f2429b48a7458990.

Sneh1999 commented on espresso/submitter/multi_worker_queue_espresso_submitter.go:
```diff
@@ -0,0 +1,1175 @@
+			if response.err != nil || response.hash == nil {
+				// Handle error case
+				job := response.job
+				log.Info("Failed to submit transaction to Espresso", "error", response.err, "attempts", job.attempt, "commit", common.Hash(job.txn.Commit()))
```
here as well change to Error

zacshowa replied:
I think that an error log might be a bit much. I would expect an error log to b a lot closer to a fatal log in terms of severity than a transaction failing to be sent, which is an operation that has retries.

Ayiga replied:
I do agree with Zac in principal, but realistically I think any error encountered here would likely be informative of potentially Espresso Network issues.

Addressed in 690fdf595a55c157fb88cbf2f2429b48a7458990.

Sneh1999 commented on espresso/submitter/multi_worker_queue_espresso_submitter.go:
```diff
@@ -0,0 +1,1175 @@
+	for {
+		select {
+		case <-ctx.Done():
+			log.Info("Submit transaction response handler exiting")
```
change to warn

Ayiga replied:
Addressed in 690fdf595a55c157fb88cbf2f2429b48a7458990.

Sneh1999 commented on espresso/submitter/multi_worker_queue_espresso_submitter.go:
```diff
@@ -0,0 +1,1175 @@
+	for i := uint64(0); i < w.numTransactionIncludedWorkers; i++ {
+		worker, workerOk := <-w.transactionIncludedJobQueue
+		if !workerOk {
+			// The job queue is closed, so we can exit.
```
same here!

Ayiga replied:
Addressed in 690fdf595a55c157fb88cbf2f2429b48a7458990.

Sneh1999 commented on espresso/submitter/multi_worker_queue_espresso_submitter.go:
```diff
@@ -0,0 +1,1175 @@
+package submitter
```
This file has become overall pretty big, I wonder if we can somehow split the code?

Ayiga replied:
Addressed in 2953e17db875e6b36fcfbb70ce357ccec8b320e4 and b59e20dd78d76f35b20deb5ea403d6eb9a10a0f7.

ImJeremyHe commented on espresso/submitter/multi_worker_queue_espresso_submitter.go:
```diff
@@ -0,0 +1,1175 @@
+	// We should always have at least one CPU core available, otherwise
+	// how is this code even being run?
+	return convertToUint64WithFallback[int, uint64](2, 1)
+}
```
It doesn't call the `runtime.NumCPU`?

Ayiga replied:
I think I adjusted this for testing or something, and never came back to it.  Good catch!

Addressed in b59e20dd78d76f35b20deb5ea403d6eb9a10a0f7.

ImJeremyHe reviewed (commented):
I might have missed something since this is a fairly large PR, but how is the case of a forced node shutdown handled? From what I see, the response handler doesn’t write anything to the DB—how would recovery work in that scenario?

zacshowa commented on espresso/submitter/multi_worker_queue_espresso_submitter.go:
```diff
@@ -0,0 +1,1175 @@
+// NOTE: This worker does not observe the context cancellation, as it is
+// expected to run until it is out of work.  Instead, it relies its job
+// queue being closed by the scheduler to indicate that no more work is left
+// for it to perform, then it will exit gracefully.
```
How does this behave when it has a large backlog of work? will it attempt to finish the entire queue before the node is able to shutdown? How does this handle sudden shutdowns/panics?

Ayiga replied:
Looking through this implementation, I'm guessing the concern for `panic`s is not coming from this code, but rather from the service running this code.  If something in the larger process has `panic`ed, then that `panic` would fall under the purview of that process, and not here.  If the larger program has `panic`ed, I think we have larger concerns.

Currently the spawned workers don't inspect the passed in context to be shutdown.  This means the workers will continue to process work, provided work is coming in via their dedicated `job queue`s.  When the scheduling loop is told to shutdown, it will iterate through all of the worker job queues, and `close` them, effectively informing them there is no more work to be processed.  It will do this as soon as it detects that it's managing context has been cancelled.  It does **NOT** currently attempt to process the remaining jobs that are pending.

It should be able to process its work backlog as quickly as it is able to.  It can build up a backlog of jobs to process that is proportional to the Job queue size, which is configurable. If that Job queue is `full`, the submission to the channel will block until it is able to insert into the `queue`.

zacshowa commented on espresso/submitter/multi_worker_queue_espresso_submitter.go:
```diff
@@ -0,0 +1,1175 @@
+}
+
+// Unwrap provides the underlying error for builtin errors checking
+func (e ErrorFailedToLaunchSubmitTransactionQueueWorker) Unwrap() error {
```
This is more a question for my own understanding, but is it not possible to implement some generic version of this once? I know in rust I would be reaching for a trait here (ignoring that this would be handled completely differently in rust). I don't think there's an equivalent in go, but is it not possible to avoid duplicating this method across every type?

Ayiga replied:
You **could** use the same `Error` struct in many places eliminating the need for this sort of repeated code.  But then your error type becomes weaker as a means of detecting that specific error.  This is the case with `fmt.Errorf`, where the underlying error is still `wrapped`, and you're able to specify some text, but the resulting error that you just returned is only detectable by the text you typed, which can be changed in the future.

zacshowa commented on espresso/submitter/multi_worker_queue_espresso_submitter.go:
```diff
@@ -0,0 +1,1175 @@
+					log.Warn("Transaction not included in Espresso within the deadline", "hash", response.job.txnHash.String(), "commit", common.Hash(response.job.txn.Commit()), "elapsed", elapsed, "attempts", response.job.attempt)
+					// We have exceeded the resubmission deadline, so we, need
+					// to try and resubmit the transaction.
+					w.submitTxnsQueue <- SubmitTransactionJob{
```
I like how elegant the solution for retries is here!

zacshowa reviewed (commented):
I think the core of the logic here looks good to me. I agree with sneh about some of the code related to types or errors being moved to different files to keep the business logic separated from the boilerplate.

I have the same question as sneh about behavior across unplanned shutdowns. What part of the code helps us resume where we left off?

I also wonder if it would be beneficial for us to have insight into the state in the channels at any point for debugging

Ayiga committed:
Relocate helpful functions to util

Addresses feedback from @Sneh1999:
- https://github.com/EspressoSystems/nitro-espresso-integration/pull/700#discussion_r2279764428
- https://github.com/EspressoSystems/nitro-espresso-integration/pull/700#discussion_r2279787412

There are a few helper functions utilities and types that have been introduced for the
multi worker queue espresso submitter that do not need to apply to it exclusively.
These helper functions have been suggested to move to a utilities location for ease of
use.

This change relocates these functions to the `util` package so they can be freely
utilized where needed.

Ayiga committed:
Split file into multiple files

Based on feedback from @Sneh1999 and @zacshowa:
- https://github.com/EspressoSystems/nitro-espresso-integration/pull/700#discussion_r2279797524
- https://github.com/EspressoSystems/nitro-espresso-integration/pull/700#pullrequestreview-3084968050

With the added functionality and types introduced for multi worker queue processing
the solution has become somewhat large and can make it more difficult to navigate
the implementation.  The suggestion has been put forth to separate the large file
into separate files.

This change separates each of the primary types into their own files for isolated
references where possible.  The errors have been kept next to their place of utilization
for faster lookup.

Ayiga committed:
Update forloop to make the endPos inclusive

Based on feedback provided by @Sneh1999:
- https://github.com/EspressoSystems/nitro-espresso-integration/pull/700#discussion_r2279769374

There is an end case in the bundleTransactions function where payloads may not be
processed as expected.  The specific case is where `startPos` == `endPos`.

This change fixes the for loop to include the `endPos`.

Ayiga committed:
Modify log levels

Based on feedback from @Sneh1999 and @zacshowa:
- https://github.com/EspressoSystems/nitro-espresso-integration/pull/700#discussion_r2279791212
- https://github.com/EspressoSystems/nitro-espresso-integration/pull/700#discussion_r2279791489
- https://github.com/EspressoSystems/nitro-espresso-integration/pull/700#discussion_r2279794032
- https://github.com/EspressoSystems/nitro-espresso-integration/pull/700#discussion_r2279795365
- https://github.com/EspressoSystems/nitro-espresso-integration/pull/700#discussion_r2279795586
- https://github.com/EspressoSystems/nitro-espresso-integration/pull/700#discussion_r2279796342
- https://github.com/EspressoSystems/nitro-espresso-integration/pull/700#discussion_r2279782182
- https://github.com/EspressoSystems/nitro-espresso-integration/pull/700#discussion_r2279782298

In order to identify potential issues as they occur in production it is beneficial to be able to identify when certain events that hinder the ability for our solution to operate.
Based on this several log levels have been changed based on requests, and others
have been added.

Ayiga committed:
Fix typo

Based on feedback from @zacshowa:
https://github.com/EspressoSystems/nitro-espresso-integration/pull/700#discussion_r2252008810

The word "utilities" is used instead of "utilizes".

This change corrects the wrong word being used.

Ayiga committed:
Modify delay on job processing failure reattempts to be configurable

The goroutine workers that are dedicated to processing a single job / task at a time
have consideration for job failure with a sleep to slowly back off requests.  This delay
is currently hard-coded to a magic number of `100ms`.  This value should really be
configurable.

This change replaces these delay values with a value that is ultimately derived from
configuration.  This allows the values to be overwritten while still allowing for a
default value.

Ayiga committed:
Modify Adapter to have positions be configurable

The `NitroMessageToEspressoTransactionAdapter` currently starts in a state where it
is unable to have any pre-existing knowledge of the data that is available or has been
submitted to Espresso.  As a result, it will always attempt to sync the messages from
position `0`.

The idea is to allow for the user to be able to specify the position that we'd like to
start submitting to Espresso.  Additionally, we would like for the user to be able to
indicate which message indexes are available.

This change allows the user to specify both of these values.

ImJeremyHe committed:
Merge branch 'integration' into ts/enh/add-alternative-espresso-submitter

ImJeremyHe commented:
@Ayiga CI is failing. Could you fix it? It is just a linter issue I believe

Ayiga committed:
Fix lint issues

Ayiga commented:
> @Ayiga CI is failing. Could you fix it? It is just a linter issue I believe

Fixed in c2a12f2a5f4741633474746685902fd1e8e6ea53.

ImJeremyHe reviewed (approved):
Looks great! I think we can add a temporary flag to enable/disable this feature so that we can test it in the devnet easily

ImJeremyHe committed:
Merge branch 'integration' into ts/enh/add-alternative-espresso-submitter

ImJeremyHe committed:
Merge branch 'integration' into ts/enh/add-alternative-espresso-submitter
